compiler/src/iree/compiler/Preprocessing/TransformExtensions/PreprocessingExtensions.cpp

```diff
@@ -0,0 +1,397 @@
+  if (target->getNumRegions() != payload->getNumRegions()) {
+    return transformOp.emitSilenceableError() << "region count mismatch";
+  }
+  for (auto [r0, r1] :
+       llvm::zip_equal(target->getRegions(), payload->getRegions())) {
+    if (!isStructurallyEquivalentTo(r0, r1)) {
+      return transformOp.emitSilenceableError()
+             << "target op does not match specified body";
+    }
+  }
+  return DiagnosedSilenceableFailure::success();
+}
+
+// Helper to check whether two operations are equivalent up to cast
+// compatibility of their arguments (i.e. the arguments of the payload
+// can be casted to the arguments of the target).
+static DiagnosedSilenceableFailure
+compareCastCompatibleOperations(transform::TransformOpInterface transformOp,
+                                Operation *target, Operation *payload) {
+  if (target->getName() != payload->getName()) {
+    return transformOp.emitSilenceableError()
+           << "target operation name " << target->getName()
+           << " does not match payload " << payload->getName();
+  }
+
+  if (target->getAttrDictionary() != payload->getAttrDictionary()) {
+    return transformOp.emitSilenceableError()
+           << "target attribute dictionary " << target->getAttrDictionary()
+           << " does not match payload attribute dictionary "
+           << payload->getAttrDictionary();
+  }
+
+  if (target->getNumResults() != payload->getNumResults()) {
+    return transformOp.emitSilenceableError() << "result count mismatch";
+  }
+
+  if (target->getNumOperands() != payload->getNumOperands()) {
+    return transformOp.emitSilenceableError() << "operand count mismatch";
+  }
+  for (auto [targetType, payloadType] :
+       llvm::zip_equal(target->getOperandTypes(), payload->getOperandTypes())) {
+    if (auto targetTensorType = dyn_cast<RankedTensorType>(targetType)) {
+      if (!isCastableToTensorType(payloadType, targetTensorType)) {
+        return transformOp.emitSilenceableError()
+               << "operand tensor type mismatch";
+      }
+    } else if (targetType != payloadType) {
+      return transformOp.emitSilenceableError() << "operand type mismatch";
+    }
+  }
+
+  return compareOperationRegions(transformOp, target, payload);
+}
+
+DiagnosedSilenceableFailure
+IREE::transform_dialect::MatchCastCompatibleDagFromRootOp::matchOperation(
+    Operation *current, transform::TransformResults &results,
+    transform::TransformState &state) {
+  Operation *targetDagRoot = getRegion().front().back().getPrevNode();
+
+  // Reserve the list of inputs based on the number of block arguments in
+  // the operation region.
+  int64_t numInputs = getRegion().getNumArguments();
+  SmallVector<Value> inputs(numInputs, nullptr);
+
+  // Maps from target/payload op to the order in which they were first
+  // processed. This is used to verify that two uses actually point to the
+  // same node in the dag.
+  llvm::MapVector<Operation *, int64_t> targetDagOrder;
```

Comment:
I am not sure you need the index. I think you could use a `MapVector<Operation *, Operation *>` to map from target to payload op and since you are checking every operand and you are traversing bottom up, you dont need to compare index, but check that you see the same operation you saw before.
Minor detail. Ok to leave as is.

This also has a general issue of this does not account for commutativity. Good for a first start, but I think upstream CSE has some methods that can be used to compare operands while respecting commutativity which can be reused in time.

Reply:
I didn't know how to easily handle commutativity so I deferred. If CSE can get it for us that would be cool, can do in a follow up and not keep growing this first patch. Probably can rework the map at the same time then.